### PR TITLE
Deprecate indices without soft-deletes

### PR DESCRIPTION
**Pull request by dnhatn**
Soft-deletes will be enabled for all indices in 8.0. Hence, we should deprecate new indices without soft-deletes in 7.x.

### CODE REVIEW

File: docs/reference/index-modules/history-retention.asciidoc [hunk 1]
```diff
@@ -65,6 +65,9 @@ there>>. {ccr-cap} will not function if soft deletes are disabled.
   configured at index creation and only on indices created on or after 6.5.0.
   The default value is `true`.
 
+  deprecated::[7.6, Indices with soft deletes disabled are deprecated and will be rejected.
```
Comment (ywelsch) [inline docs/reference/index-modules/history-retention.asciidoc]: > and will be rejected

This sounds as if these indices will no longer be supported. Perhaps reword this to: `Creating indices with soft-deletes disabled is deprecated and will be removed in future Elasticsearch versions.`

File: qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/RecoveryIT.java [hunk 2]
```diff
@@ -294,7 +295,7 @@ public void testRecoveryWithSoftDeletes() throws Exception {
                 .put(INDEX_DELAYED_NODE_LEFT_TIMEOUT_SETTING.getKey(), "100ms")
                 .put(SETTING_ALLOCATION_MAX_RETRY.getKey(), "0"); // fail faster
             if (randomBoolean()) {
-                settings.put(IndexSettings.INDEX_SOFT_DELETES_SETTING.getKey(), true);
+                settings.put(IndexSettings.INDEX_SOFT_DELETES_SETTING.getKey(), randomBoolean());
```
Comment (ywelsch) [inline qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/RecoveryIT.java]: why change this? Should the test be named differently in that case (e.g. just `testRecovery`)?

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java [hunk 3]
```diff
@@ -434,6 +437,12 @@ static Settings aggregateIndexSettings(ClusterState currentState, CreateIndexClu
          * that will be used to create this index.
          */
         MetaDataCreateIndexService.checkShardLimit(indexSettings, currentState);
+        if (indexSettings.getAsBoolean(IndexSettings.INDEX_SOFT_DELETES_SETTING.getKey(), true) == false) {
+            DEPRECATION_LOGGER.deprecatedAndMaybeLog("soft_deletes_disabled",
+                "Index with soft deletes disabled is deprecated and will be rejected. " +
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java]: also adapt message here

**Review by ywelsch — CHANGES_REQUESTED** [review]

File: qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/RecoveryIT.java (hunk 2 above)
Comment (dnhatn) [reply qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/RecoveryIT.java]: We introduced this test when soft-deletes was disabled by default. We should invert the test condition as soft-deletes is enabled by default now. I renamed it in https://github.com/elastic/elasticsearch/pull/50502/commits/893d28ea7ad030c29390160163f952e4ea421a67.

File: docs/reference/index-modules/history-retention.asciidoc (hunk 1 above)
Comment (dnhatn) [reply docs/reference/index-modules/history-retention.asciidoc]: ++. Adjusted in https://github.com/elastic/elasticsearch/pull/50502/commits/867050d34b0d56783e11f6435afa8d1809bc157c.

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java (hunk 3 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java]: Adjusted in https://github.com/elastic/elasticsearch/pull/50502/commits/867050d34b0d56783e11f6435afa8d1809bc157c.

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java [hunk 4]
```diff
@@ -434,6 +437,11 @@ static Settings aggregateIndexSettings(ClusterState currentState, CreateIndexClu
          * that will be used to create this index.
          */
         MetaDataCreateIndexService.checkShardLimit(indexSettings, currentState);
+        if (indexSettings.getAsBoolean(IndexSettings.INDEX_SOFT_DELETES_SETTING.getKey(), true) == false) {
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java]: Do we want to completely disallow setting this setting in 8.x? If so, should we always emit a warning, even if it's explicitly set to true?

**Review by ywelsch — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java (hunk 4 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java]: Users might be confused about the deprecation when creating indices for CCR in a mixed cluster between 7.x and 6.8 because they need to explicitly set the soft_deletes setting.

**Comment by dnhatn** [comment]
@ywelsch Thanks for reviewing.

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java (hunk 4 above)
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java]: If mixed clusters with 6.8 is a worry, you can just check here that all nodes are 7.0+.
With the approach here, we will still have to support the setting for the full duration of 8.x (i.e. allow people to explicitly set soft-deletes true).

Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java]: ++. I will make this change in a follow-up.
